bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py

```diff
@@ -106,6 +106,80 @@ def _(operation: Snapshot, range_id: int) -> None:
         operation.args.append(lazy_options)  # Append RSnapshotOptions
 
 
+@singledispatch
+def generate_computation_graph(distrdf_node: Node, previous_node, range_id):
+    """
+    Generates the RDF computation graph by recursively retrieving
+    information from the DistRDF nodes.
+
+    Args:
+        distrdf_node: The current DistRDF node in the computation graph. In the
+            first recursive state this is the DistRDF headnode.
+        previous_node: The node in the RDF computation graph on which
+            the operation of the current recursive state is called. In the
+            first recursive state, this corresponds to the RDataFrame
+            object that will be processed. Specifically, if the head node
+            of the computation graph is an EmptySourceHeadNode, then the
+            first current node will actually be the result of a call to the
+            Range operation. If the head node is a TreeHeadNode then the
+            node will be an actual RDataFrame. Successive recursive states
+            will receive the result of an RDF operation call
+            (e.g. Histo1D, Count).
+        range_id (int): The id of the current range. Needed to assign a
+            file name to a partial Snapshot if it was requested.
+
+    Returns:
+        list: List of actions of the computation graph to be triggered. Each
+        element is some kind of promise of a result (usually an
+        RResultPtr). Exceptions are the 'AsNumpy' operation for which an
+        'AsNumpyResult' is returned and the 'Snapshot' operation for which a
+        'SnapshotResult' is returned.
+    """
+    future_results = []
+
+    # Execute the current operation using the output of the previous node
+    RDFOperation = getattr(previous_node, distrdf_node.operation.name)
+    operation = distrdf_node.operation
+    _make_op_lazy_if_needed(operation, range_id)
+    pyroot_node = RDFOperation(*operation.args, **operation.kwargs)
+
+    # The result is a pyroot object which is stored together with
+    # the DistRDF node. This binds the pyroot object lifetime to the
+    # DistRDF node, so both nodes will be kept alive as long as there
+    # is a valid reference pointing to the DistRDF node.
+    distrdf_node.pyroot_node = pyroot_node
+
+    # Set the next `previous_node` input argument to the `pyroot_node`
+    # we just retrieved
+    previous_node = pyroot_node
+
+    append_node_to_results(operation, pyroot_node, future_results)
+
+    for child_node in distrdf_node.children:
+        # Recurse through children and get their output
+        prev_results = generate_computation_graph(child_node, previous_node, range_id)
+
+        # Attach the output of the children node
+        future_results.extend(prev_results)
+
+    return future_results
+
+
+@generate_computation_graph.register(HeadNode)
+def _(distrdf_node, previous_node, range_id):
+
+    future_results = []
+
+    for child_node in distrdf_node.children:
```

Comment:
I am not sure this change is beneficial since it incurs in code repetition?

Reply:
Indeed, this can be done better. I'll update the commits with a better separation of roles in the functions